sktree/tree/_oblique_splitter.pyx

```diff
@@ -596,3 +620,205 @@ cdef class BestObliqueSplitter(ObliqueSplitter):
+    cdef int node_split(
+        self,
+        double impurity,
+        SplitRecord* split,
+        SIZE_t* n_constant_features
+    ) except -1 nogil:
+        """Find the best_split split on node samples[start:end]
+
+        Returns -1 in case of failure to allocate memory (and raise MemoryError)
+        or 0 otherwise.
+        """
+        # typecast the pointer to an ObliqueSplitRecord
+        cdef ObliqueSplitRecord* oblique_split = <ObliqueSplitRecord*>(split)
+
+        # Draw random splits and pick the best_split
+        cdef SIZE_t[::1] samples = self.samples
+        cdef SIZE_t start = self.start
+        cdef SIZE_t end = self.end
+        cdef UINT32_t* random_state = &self.rand_r_state
+
+        # pointer array to store feature values to split on
+        cdef DTYPE_t[::1]  feature_values = self.feature_values
+        cdef SIZE_t max_features = self.max_features
+        cdef SIZE_t min_samples_leaf = self.min_samples_leaf
+        cdef double min_weight_leaf = self.min_weight_leaf
+
+        # keep track of split record for current_split node and the best_split split
+        # found among the sampled projection vectors
+        cdef ObliqueSplitRecord best_split, current_split
+        cdef double current_proxy_improvement = -INFINITY
+        cdef double best_proxy_improvement = -INFINITY
+
+        cdef SIZE_t feat_i, p       # index over computed features and start/end
+        cdef SIZE_t partition_end
+        cdef DTYPE_t temp_d         # to compute a projection feature value
+        cdef DTYPE_t min_feature_value
+        cdef DTYPE_t max_feature_value
+
+        # instantiate the split records
+        _init_split(&best_split, end)
+
+        # Sample the projection matrix
+        self.sample_proj_mat(self.proj_mat_weights, self.proj_mat_indices)
+
+        # For every vector in the projection matrix
+        for feat_i in range(max_features):
+            # Projection vector has no nonzeros
+            if self.proj_mat_weights[feat_i].empty():
+                continue
+
+            # XXX: 'feature' is not actually used in oblique split records
+            # Just indicates which split was sampled
+            current_split.feature = feat_i
+            current_split.proj_vec_weights = &self.proj_mat_weights[feat_i]
+            current_split.proj_vec_indices = &self.proj_mat_indices[feat_i]
+
+            # Compute linear combination of features and then
+            # sort samples according to the feature values.
+            self.compute_features_over_samples(
+                start,
+                end,
+                samples,
+                feature_values,
+                &self.proj_mat_weights[feat_i],
+                &self.proj_mat_indices[feat_i]
+            )
+
+            # find min, max of the feature_values
+            self.find_min_max(feature_values, &min_feature_value, &max_feature_value)
+
+            # Draw a random threshold
```

Comment:
```suggestion
            # XXX: we can still add some logic here to keep track of constant features if they exist
            # Draw a random threshold
```

Reply:
In the sklearn version of Splitter class, the features are organized so constants are at the front and non-constants are at the back. But does it make sense to do this in Oblique trees? My guess was no, in which case, how do we use the variable `n_constant_features`?

Reply:
We don't use it rn but in the future we can.

Reply:
This is just to add an in-line comment, so we remember to do so in the future.

Basically if we run OF and at any point we randomly choose only one feature and realize it is a constant feature, we can note that.